src/Expander/Syntax.hs

```diff
@@ -59,38 +64,72 @@ mustBeModName (Syntax (Stx scs srcloc (Id "kernel"))) =
 mustBeModName other = throwError (NotModName other)
 
 
-class FixedLengthList a where
-  mustHaveEntries :: Syntax -> Expand (Stx a)
-
-instance FixedLengthList () where
-  mustHaveEntries (Syntax (Stx scs srcloc (List []))) = return (Stx scs srcloc ())
-  mustHaveEntries other = throwError (NotRightLength 0 other)
-
-instance FixedLengthList Syntax where
-  mustHaveEntries (Syntax (Stx scs srcloc (List [x]))) = return (Stx scs srcloc x)
-  mustHaveEntries other = throwError (NotRightLength 1 other)
-
-instance (a ~ Syntax, b ~ Syntax) => FixedLengthList (a, b) where
-  mustHaveEntries (Syntax (Stx scs srcloc (List [x, y]))) = return (Stx scs srcloc (x, y))
-  mustHaveEntries other = throwError (NotRightLength 2 other)
-
-instance (a ~ Syntax, b ~ Syntax, c ~ Syntax) => FixedLengthList (a, b, c) where
-  mustHaveEntries (Syntax (Stx scs srcloc (List [x, y, z]))) = return (Stx scs srcloc (x, y, z))
-  mustHaveEntries other = throwError (NotRightLength 3 other)
-
-instance (a ~ Syntax, b ~ Syntax, c ~ Syntax, d ~ Syntax) => FixedLengthList (a, b, c, d) where
-  mustHaveEntries (Syntax (Stx scs srcloc (List [w, x, y, z]))) = return (Stx scs srcloc (w, x, y, z))
-  mustHaveEntries other = throwError (NotRightLength 4 other)
-
-instance (a ~ Syntax, b ~ Syntax, c ~ Syntax, d ~ Syntax, e ~ Syntax) => FixedLengthList (a, b, c, d, e) where
-  mustHaveEntries (Syntax (Stx scs srcloc (List [v, w, x, y, z]))) =
-    return (Stx scs srcloc (v, w, x, y, z))
-  mustHaveEntries other = throwError (NotRightLength 5 other)
-
-instance (a ~ Syntax) => FixedLengthList [a] where
-  mustHaveEntries (Syntax (Stx scs srcloc (List xs))) =
-    return (Stx scs srcloc xs)
-  mustHaveEntries other = throwError (NotList other)
+mustHaveEntries
+  :: ( FixedLengthList item r
+     , item ~ Syntax
+     )
+  => Syntax -> Expand (Stx r)
+mustHaveEntries stx@(Syntax (Stx scs srcloc (List xs))) = do
+  case checkLength xs of
+    Right r -> do
+      pure (Stx scs srcloc r)
+    Left lengths -> do
+      throwError (NotRightLength lengths stx)
+mustHaveEntries other = do
+  throwError (NotList other)
+
+class FixedLengthList item r where
+  checkLength :: [item] -> Either [Natural] r
+
+instance ( FixedLengthList e1 a
+         , FixedLengthList e2 b
+         , e1 ~ e2
```

Comment:
hmm, I don't think that's right, I don't need to use two different type variables `e1` and `e2` here.

Reply:
Fixed in 2ca2a30a4a731068974fb7c21f75d424b0a69bea, and this change also removes the need for `UndecidableInstances` in this file.

Reply:
:confetti_ball: